contracts/shapella-upgrade/ShapellaUpgradeTemplate.sol

```diff
@@ -0,0 +1,794 @@
+        _transferOZAdminFromThisToAgent(_hashConsensusForAccountingOracle);
+        _transferOZAdminFromThisToAgent(_burner());
+        _transferOZAdminFromThisToAgent(_stakingRouter());
+        _transferOZAdminFromThisToAgent(_accountingOracle());
+        _transferOZAdminFromThisToAgent(_validatorsExitBusOracle());
+        _transferOZAdminFromThisToAgent(_withdrawalQueue());
+
+        _locator.proxy__changeAdmin(_agent);
+        _stakingRouter().proxy__changeAdmin(_agent);
+        _accountingOracle().proxy__changeAdmin(_agent);
+        _validatorsExitBusOracle().proxy__changeAdmin(_agent);
+        _withdrawalQueue().proxy__changeAdmin(_agent);
+
+        _depositSecurityModule().setOwner(_agent);
+    }
+
+    function _assertUpgradeIsFinishedCorrectly() internal view {
+        _checkContractVersions();
+
+        _assertAdminsOfProxies(_agent);
+        _assertProxyOZAccessControlContractsAdmin(_agent);
+        _assertNonProxyOZAccessControlContractsAdmin(_agent);
+        _assertOracleDaemonConfigRoles();
+        _assertOracleReportSanityCheckerRoles();
+
+        _assertGateSealSealables();
+
+        _assertZeroOZRoleHolders(_withdrawalQueue(), _withdrawalQueue().RESUME_ROLE());
+        _assertSingleOZRoleHolder(_withdrawalQueue(), _withdrawalQueue().PAUSE_ROLE(), _gateSeal);
+        _assertSingleOZRoleHolder(_withdrawalQueue(), _withdrawalQueue().FINALIZE_ROLE(), address(_lido()));
+        _assertSingleOZRoleHolder(_withdrawalQueue(), _withdrawalQueue().ORACLE_ROLE(), address(_accountingOracle()));
+        _assertZeroOZRoleHolders(_stakingRouter(), _stakingRouter().STAKING_MODULE_RESUME_ROLE());
+        _assertSingleOZRoleHolder(_stakingRouter(), _stakingRouter().STAKING_MODULE_PAUSE_ROLE(), address(_depositSecurityModule()));
+        _assertSingleOZRoleHolder(_stakingRouter(), _stakingRouter().REPORT_EXITED_VALIDATORS_ROLE(), address(_accountingOracle()));
+        _assertSingleOZRoleHolder(_stakingRouter(), _stakingRouter().REPORT_REWARDS_MINTED_ROLE(), address(_lido()));
+        _assertTwoOZRoleHolders(_burner(), _burner().REQUEST_BURN_SHARES_ROLE(), address(_lido()), address(_nodeOperatorsRegistry));
+        _assertZeroOZRoleHolders(_validatorsExitBusOracle(), _validatorsExitBusOracle().RESUME_ROLE());
+        _assertSingleOZRoleHolder(_validatorsExitBusOracle(), _validatorsExitBusOracle().PAUSE_ROLE(), _gateSeal);
+
+        if (_depositSecurityModule().getOwner() != _agent) revert WrongDsmOwner();
+
+        _assertCorrectOracleAndConsensusContractsBinding(_accountingOracle(), _hashConsensusForAccountingOracle);
+        _assertCorrectOracleAndConsensusContractsBinding(_validatorsExitBusOracle(), _hashConsensusForValidatorsExitBusOracle);
+
+        if (_withdrawalQueue().isPaused()) revert WQNotResumed();
+        if (_validatorsExitBusOracle().isPaused()) revert VEBONotResumed();
+
+        if (!_stakingRouter().hasStakingModule(1) || _stakingRouter().hasStakingModule(2)) {
```

Comment:
not a blocker
seems worth to check it's the correct stakingModule

Reply:
done